asapdiscovery-data/asapdiscovery/data/state_expanders/protomer_expander.py

```diff
@@ -0,0 +1,208 @@
+    )
+
+    def _create_cmd(self, *programs: str) -> str:
+        """
+        Create a command which can be used to call some SCHRODINGER software
+        Returns
+        -------
+            The string which can be passed to subprocess to call epik
+        """
+        # create a path to epik
+        schrodinger_folder = os.getenv("SCHRODINGER")
+        epik = os.path.join(schrodinger_folder, *programs)
+        return epik
+
+    def _provenance(self) -> dict[str, str]:
+        """
+        Run epik to get the version info.
+        Returns
+        -------
+            The version of epik used.
+        """
+        epik_cmd = self._create_cmd("epik")
+        # call epik to get the version info
+        output = subprocess.check_output([epik_cmd, "-v"])
+        for line in output.decode("utf-8").split("\n"):
+            if "Epik version" in line:
+                version = line.split()[-1]
+                break
+        else:
+            version = "unknown"
+
+        return {
+            "epik": version,
+        }
+
+    def _prepare_ligands(self, ligands: list[Ligand]):
+        """
+        Convert the list of Ligands to a SCHRODINGER mae file before running with Epik.
+        """
+        oe_ligands = [ligand.to_oemol() for ligand in ligands]
+        save_openeye_sdfs(oe_ligands, "input.sdf")
+        convert_cmd = self._create_cmd("utilities", "structconvert")
+        with open("structconvert.log", "w") as log:
+            subprocess.run(
+                convert_cmd + " input.sdf input.mae",
+                shell=True,
+                stdout=log,
+                stderr=log,
+                check=True,
+            )
+
+    def _extract_ligands(self) -> list[Ligand]:
+        """
+        Extract the state expanded ligands from the Epik output file.
+        Returns
+        -------
+            A list of expanded state ligands.
+        """
+        convert_cmd = self._create_cmd("utilities", "structconvert")
+        with open("structconvert.log", "w") as log:
+            subprocess.run(
+                convert_cmd + " output.mae output.sdf",
+                shell=True,
+                stdout=log,
+                stderr=log,
+                check=True,
+            )
+        oe_mols = load_openeye_sdfs(sdf_fn="output.sdf")
+        # parse into ligand objects
+        expanded_ligands = [Ligand.from_oemol(oemol) for oemol in oe_mols]
+        return expanded_ligands
+
+    def _call_epik(self):
+        """Call Epik on the local ligands file."""
+        import numpy as np
+
+        epik_command = self._create_cmd("epik")
+        min_population = np.exp(-6)
+        epik_command += f" -WAIT -ms 16 -ph {self.ph} -p {min_population} -imae input.mae -omae output.mae"
+        with open("epik_log.log", "w") as log:
+            subprocess.run(epik_command, shell=True, stdout=log, stderr=log, check=True)
+
+    def _expand(self, ligands: list[Ligand]) -> list[Ligand]:
```

Comment:
Nicely done, you have made the best of a bad situation lol.

Reply:
Thanks its still not great and a pain to test so I see this being a problem in future :(